### PR TITLE
Add ability to translate between SONiC interface name and vendor alias

### PR DESCRIPTION
**Pull request by chaliu**
This is a patch to enable alias translation. If the gNMI request uses vendor aliases for interface names, it first translate the vendor alias into SONiC interface name, then query Redis database using SONiC interface name. When it comes to response, if the gNMI request uses vendor aliases, response message also uses vendor aliases; if the gNMI request uses SONiC interface name, the response message uses SONiC interface name as well.

### CODE REVIEW

**Comment by chaliu** [comment]
The unit test cases for alias translation is added. Please review the code.

**Comment by jipanyang** [comment]
@chaliu thanks for the updates. Since we don't have auto test check available for this repo yet, could you paste your local test result here, for test cases under both gnmi_server and dialout/dialout_client?

**Comment by chaliu** [comment]
Hi, Jipan,

I haven’t tested dialout module.

For gnmi_server test, I found that original code fails beginning from first “TestGnmiSubscribe” test, as you can see below
=== RUN   TestGnmiGet
=== RUN   TestGnmiGet/Test_non-existing_path_Target
=== RUN   TestGnmiGet/Test_empty_path_target
=== RUN   TestGnmiGet/Get_valid_but_non-existing_node
=== RUN   TestGnmiGet/Get_COUNTERS_PORT_NAME_MAP
=== RUN   TestGnmiGet/get_COUNTERS:Ethernet68
=== RUN   TestGnmiGet/get_COUNTERS:Ethernet68_SAI_PORT_STAT_PFC_7_RX_PKTS
=== RUN   TestGnmiGet/get_COUNTERS:Ethernet*
=== RUN   TestGnmiGet/get_COUNTERS:Ethernet*_SAI_PORT_STAT_PFC_7_RX_PKTS
--- PASS: TestGnmiGet (1.58s)
    --- PASS: TestGnmiGet/Test_non-existing_path_Target (0.02s)
    --- PASS: TestGnmiGet/Test_empty_path_target (0.00s)
    --- PASS: TestGnmiGet/Get_valid_but_non-existing_node (0.00s)
    --- PASS: TestGnmiGet/Get_COUNTERS_PORT_NAME_MAP (0.00s)
    --- PASS: TestGnmiGet/get_COUNTERS:Ethernet68 (0.00s)
    --- PASS: TestGnmiGet/get_COUNTERS:Ethernet68_SAI_PORT_STAT_PFC_7_RX_PKTS (0.00s)
    --- PASS: TestGnmiGet/get_COUNTERS:Ethernet* (0.00s)
    --- PASS: TestGnmiGet/get_COUNTERS:Ethernet*_SAI_PORT_STAT_PFC_7_RX_PKTS (0.00s)
=== RUN   TestGnmiSubscribe
=== RUN   TestGnmiSubscribe/stream_query_for_table_COUNTERS_PORT_NAME_MAP_with_new_test_field_field
=== RUN   TestGnmiSubscribe/stream_query_for_table_key_Ethernet68_with_new_test_field_field
=== RUN   TestGnmiSubscribe/stream_query_for_COUNTERS/Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_with_update_of_filed_value
=== RUN   TestGnmiSubscribe/stream_query_for_table_key_Ethernet*_with_new_test_field_field_on_Ethernet68
=== RUN   TestGnmiSubscribe/stream_query_for_table_key_Ethernet*/SAI_PORT_STAT_PFC_7_RX_PKTS_with_field_value_update
=== RUN   TestGnmiSubscribe/poll_query_for_table_COUNTERS_PORT_NAME_MAP_with_new_field_test_field
=== RUN   TestGnmiSubscribe/poll_query_for_table_COUNTERS_PORT_NAME_MAP_with_test_field_delete
=== RUN   TestGnmiSubscribe/poll_query_for_COUNTERS/Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_with_field_value_change
=== RUN   TestGnmiSubscribe/poll_query_for_table_key_Ethernet*_with_Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_field_value_change
=== RUN   TestGnmiSubscribe/poll_query_for_table_key_field_Ethernet*/SAI_PORT_STAT_PFC_7_RX_PKTS_with_Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_field_value_change
=== RUN   TestGnmiSubscribe/poll_query_for_COUNTERS/Ethernet*/Queues
=== RUN   TestGnmiSubscribe/poll_query_for_COUNTERS/Ethernet68/Queues_with_field_value_change
--- FAIL: TestGnmiSubscribe (41.50s)
    --- FAIL: TestGnmiSubscribe/stream_query_for_table_COUNTERS_PORT_NAME_MAP_with_new_test_field_field (2.01s)


I added unit test in the “TestGnmiGet” section, and they passed the test:
=== RUN   TestGnmiGet
=== RUN   TestGnmiGet/Test_non-existing_path_Target
=== RUN   TestGnmiGet/Test_empty_path_target
=== RUN   TestGnmiGet/Get_valid_but_non-existing_node
=== RUN   TestGnmiGet/Get_COUNTERS_PORT_NAME_MAP
=== RUN   TestGnmiGet/get_COUNTERS:Ethernet68
=== RUN   TestGnmiGet/get_COUNTERS:Ethernet68_SAI_PORT_STAT_PFC_7_RX_PKTS
=== RUN   TestGnmiGet/use_vendor_alias:_get_COUNTERS:Ethernet68/1
=== RUN   TestGnmiGet/use_vendor_alias:_get_COUNTERS:Ethernet68/1_SAI_PORT_STAT_PFC_7_RX_PKTS
=== RUN   TestGnmiGet/get_COUNTERS:Ethernet*
=== RUN   TestGnmiGet/get_COUNTERS:Ethernet*_SAI_PORT_STAT_PFC_7_RX_PKTS
--- PASS: TestGnmiGet (2.57s)
    --- PASS: TestGnmiGet/Test_non-existing_path_Target (0.01s)
    --- PASS: TestGnmiGet/Test_empty_path_target (0.00s)
    --- PASS: TestGnmiGet/Get_valid_but_non-existing_node (0.00s)
    --- PASS: TestGnmiGet/Get_COUNTERS_PORT_NAME_MAP (0.00s)
    --- PASS: TestGnmiGet/get_COUNTERS:Ethernet68 (0.00s)
    --- PASS: TestGnmiGet/get_COUNTERS:Ethernet68_SAI_PORT_STAT_PFC_7_RX_PKTS (0.00s)
    --- PASS: TestGnmiGet/use_vendor_alias:_get_COUNTERS:Ethernet68/1 (0.00s)
    --- PASS: TestGnmiGet/use_vendor_alias:_get_COUNTERS:Ethernet68/1_SAI_PORT_STAT_PFC_7_RX_PKTS (0.00s)
    --- PASS: TestGnmiGet/get_COUNTERS:Ethernet* (0.00s)
    --- PASS: TestGnmiGet/get_COUNTERS:Ethernet*_SAI_PORT_STAT_PFC_7_RX_PKTS (0.00s)

Need your help or Liuqu to fix the Subscribe bug in the original code.

Thanks
Chang

**Comment by chaliu** [comment]
I added unit test cases in dial-out mode. 

Again, the original code fails in the first test and passes the second test, as shown:
=== RUN   TestGNMIDialOutPublish
=== RUN   TestGNMIDialOutPublish/DialOut_to_first_collector_in_stream_mode_and_synced
=== RUN   TestGNMIDialOutPublish/DialOut_to_second_collector_in_stream_mode_upon_failure_of_first_collector
--- FAIL: TestGNMIDialOutPublish (19.83s)
    --- FAIL: TestGNMIDialOutPublish/DialOut_to_first_collector_in_stream_mode_and_synced (0.51s)
    --- PASS: TestGNMIDialOutPublish/DialOut_to_second_collector_in_stream_mode_upon_failure_of_first_collector (7.51s)



After alias translation is added, I got the same result: the first test case fails and the second passes:
=== RUN   TestGNMIDialOutPublish
=== RUN   TestGNMIDialOutPublish/DialOut_to_first_collector_in_stream_mode_and_synced
=== RUN   TestGNMIDialOutPublish/DialOut_to_second_collector_in_stream_mode_upon_failure_of_first_collector
--- FAIL: TestGNMIDialOutPublish (19.03s)
    --- FAIL: TestGNMIDialOutPublish/DialOut_to_first_collector_in_stream_mode_and_synced (0.50s)
    --- PASS: TestGNMIDialOutPublish/DialOut_to_second_collector_in_stream_mode_upon_failure_of_first_collector (7.51s)


Need Jipan or Liuqu's help to fix the bug in the original code.

Thanks
Chang

**Comment by jipanyang** [comment]
The unit test works for me.  maybe you could try "go get -u all" to get the latest package then run it again.

But I hit the problem that go test didn't exit automatically after test finish, not sure if it is the issue with latest golang.

 

```
US-70998M:sonic-telemetry jipanyang$ git status
On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working tree clean
US-70998M:sonic-telemetry jipanyang$ git log -n 2
commit 174649414301082dad363196d069a34b503e07ae
Author: Jipan Yang <jipan.yang@alibaba-inc.com>
Date:   Tue Aug 21 02:52:02 2018 -0700

    Add debian package support (#11)
    
    Signed-off-by: Jipan Yang <jipan.yang@alibaba-inc.com>

commit a1407aa919737c77b9876f2e80ef1a2516ec7071
Author: Jipan Yang <jipan.yang@alibaba-inc.com>
Date:   Sat Mar 24 18:47:07 2018 -0700

    Add basic installing and running instructions in README (#10)
    
    Signed-off-by: Jipan Yang <jipan.yang@alibaba-inc.com>
US-70998M:sonic-telemetry jipanyang$ go test -v ./dialout/dialout_client/ 
=== RUN   TestGNMIDialOutPublish
=== RUN   TestGNMIDialOutPublish/DialOut_to_first_collector_in_stream_mode_and_synced
=== RUN   TestGNMIDialOutPublish/DialOut_to_second_collector_in_stream_mode_upon_failure_of_first_collector
--- PASS: TestGNMIDialOutPublish (25.98s)
    --- PASS: TestGNMIDialOutPublish/DialOut_to_first_collector_in_stream_mode_and_synced (0.53s)
    --- PASS: TestGNMIDialOutPublish/DialOut_to_second_collector_in_stream_mode_upon_failure_of_first_collector (7.53s)
PASS
ok  	sonic-telemetry/dialout/dialout_client	26.007s






^C^Z
[4]+  Stopped                 go test -v ./dialout/dialout_client/
US-70998M:sonic-telemetry jipanyang$ go test -v ./gnmi_server/
=== RUN   TestGnmiGet
=== RUN   TestGnmiGet/Test_non-existing_path_Target
=== RUN   TestGnmiGet/Test_empty_path_target
=== RUN   TestGnmiGet/Get_valid_but_non-existing_node
=== RUN   TestGnmiGet/Get_COUNTERS_PORT_NAME_MAP
=== RUN   TestGnmiGet/get_COUNTERS:Ethernet68
=== RUN   TestGnmiGet/get_COUNTERS:Ethernet68_SAI_PORT_STAT_PFC_7_RX_PKTS
=== RUN   TestGnmiGet/get_COUNTERS:Ethernet*
=== RUN   TestGnmiGet/get_COUNTERS:Ethernet*_SAI_PORT_STAT_PFC_7_RX_PKTS
--- PASS: TestGnmiGet (4.93s)
    --- PASS: TestGnmiGet/Test_non-existing_path_Target (0.10s)
    --- PASS: TestGnmiGet/Test_empty_path_target (0.00s)
    --- PASS: TestGnmiGet/Get_valid_but_non-existing_node (0.00s)
    --- PASS: TestGnmiGet/Get_COUNTERS_PORT_NAME_MAP (0.00s)
    --- PASS: TestGnmiGet/get_COUNTERS:Ethernet68 (0.00s)
    --- PASS: TestGnmiGet/get_COUNTERS:Ethernet68_SAI_PORT_STAT_PFC_7_RX_PKTS (0.00s)
    --- PASS: TestGnmiGet/get_COUNTERS:Ethernet* (0.01s)
    --- PASS: TestGnmiGet/get_COUNTERS:Ethernet*_SAI_PORT_STAT_PFC_7_RX_PKTS (0.01s)
=== RUN   TestGnmiSubscribe
=== RUN   TestGnmiSubscribe/stream_query_for_table_COUNTERS_PORT_NAME_MAP_with_new_test_field_field
=== RUN   TestGnmiSubscribe/stream_query_for_table_key_Ethernet68_with_new_test_field_field
=== RUN   TestGnmiSubscribe/stream_query_for_COUNTERS/Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_with_update_of_filed_value
12077:M 23 Aug 19:58:50.020 * 100 changes in 300 seconds. Saving...
12077:M 23 Aug 19:58:50.021 * Background saving started by pid 12201
12201:C 23 Aug 19:58:50.026 * DB saved on disk
12077:M 23 Aug 19:58:50.125 * Background saving terminated with success
=== RUN   TestGnmiSubscribe/stream_query_for_table_key_Ethernet*_with_new_test_field_field_on_Ethernet68
=== RUN   TestGnmiSubscribe/stream_query_for_table_key_Ethernet*/SAI_PORT_STAT_PFC_7_RX_PKTS_with_field_value_update
=== RUN   TestGnmiSubscribe/poll_query_for_table_COUNTERS_PORT_NAME_MAP_with_new_field_test_field
=== RUN   TestGnmiSubscribe/poll_query_for_table_COUNTERS_PORT_NAME_MAP_with_test_field_delete
=== RUN   TestGnmiSubscribe/poll_query_for_COUNTERS/Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_with_field_value_change
=== RUN   TestGnmiSubscribe/poll_query_for_table_key_Ethernet*_with_Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_field_value_change
=== RUN   TestGnmiSubscribe/poll_query_for_table_key_field_Ethernet*/SAI_PORT_STAT_PFC_7_RX_PKTS_with_Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_field_value_change
=== RUN   TestGnmiSubscribe/poll_query_for_COUNTERS/Ethernet*/Queues
=== RUN   TestGnmiSubscribe/poll_query_for_COUNTERS/Ethernet68/Queues_with_field_value_change
--- PASS: TestGnmiSubscribe (41.77s)
    --- PASS: TestGnmiSubscribe/stream_query_for_table_COUNTERS_PORT_NAME_MAP_with_new_test_field_field (2.01s)
    --- PASS: TestGnmiSubscribe/stream_query_for_table_key_Ethernet68_with_new_test_field_field (3.01s)
    --- PASS: TestGnmiSubscribe/stream_query_for_COUNTERS/Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_with_update_of_filed_value (3.00s)
    --- PASS: TestGnmiSubscribe/stream_query_for_table_key_Ethernet*_with_new_test_field_field_on_Ethernet68 (3.01s)
    --- PASS: TestGnmiSubscribe/stream_query_for_table_key_Ethernet*/SAI_PORT_STAT_PFC_7_RX_PKTS_with_field_value_update (2.01s)
    --- PASS: TestGnmiSubscribe/poll_query_for_table_COUNTERS_PORT_NAME_MAP_with_new_field_test_field (2.02s)
    --- PASS: TestGnmiSubscribe/poll_query_for_table_COUNTERS_PORT_NAME_MAP_with_test_field_delete (2.01s)
    --- PASS: TestGnmiSubscribe/poll_query_for_COUNTERS/Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_with_field_value_change (2.01s)
    --- PASS: TestGnmiSubscribe/poll_query_for_table_key_Ethernet*_with_Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_field_value_change (2.03s)
    --- PASS: TestGnmiSubscribe/poll_query_for_table_key_field_Ethernet*/SAI_PORT_STAT_PFC_7_RX_PKTS_with_Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_field_value_change (2.03s)
    --- PASS: TestGnmiSubscribe/poll_query_for_COUNTERS/Ethernet*/Queues (1.12s)
    --- PASS: TestGnmiSubscribe/poll_query_for_COUNTERS/Ethernet68/Queues_with_field_value_change (2.01s)
PASS
ok  	sonic-telemetry/gnmi_server	46.733s
```

```
US-70998M:sonic-telemetry jipanyang$ go version
go version go1.10.3 darwin/amd64
```

**Comment by chaliu** [comment]
@jipanyang I fixed the bugs in the unit tests, now everything passes. Please review the code.

```
cchliu@cchliu-vm:~/Work/go/src/test/sonic-telemetry-backup$ go test -v ./gnmi_server/
=== RUN   TestGnmiGet
=== RUN   TestGnmiGet/Test_non-existing_path_Target
=== RUN   TestGnmiGet/Test_empty_path_target
=== RUN   TestGnmiGet/Get_valid_but_non-existing_node
=== RUN   TestGnmiGet/Get_COUNTERS_PORT_NAME_MAP
=== RUN   TestGnmiGet/get_COUNTERS:Ethernet68
=== RUN   TestGnmiGet/get_COUNTERS:Ethernet68_SAI_PORT_STAT_PFC_7_RX_PKTS
=== RUN   TestGnmiGet/use_vendor_alias:_get_COUNTERS:Ethernet68/1
=== RUN   TestGnmiGet/use_vendor_alias:_get_COUNTERS:Ethernet68/1_SAI_PORT_STAT_PFC_7_RX_PKTS
=== RUN   TestGnmiGet/get_COUNTERS:Ethernet*
=== RUN   TestGnmiGet/get_COUNTERS:Ethernet*_SAI_PORT_STAT_PFC_7_RX_PKTS
--- PASS: TestGnmiGet (0.75s)
    --- PASS: TestGnmiGet/Test_non-existing_path_Target (0.02s)
    --- PASS: TestGnmiGet/Test_empty_path_target (0.00s)
    --- PASS: TestGnmiGet/Get_valid_but_non-existing_node (0.00s)
    --- PASS: TestGnmiGet/Get_COUNTERS_PORT_NAME_MAP (0.00s)
    --- PASS: TestGnmiGet/get_COUNTERS:Ethernet68 (0.00s)
    --- PASS: TestGnmiGet/get_COUNTERS:Ethernet68_SAI_PORT_STAT_PFC_7_RX_PKTS (0.00s)
    --- PASS: TestGnmiGet/use_vendor_alias:_get_COUNTERS:Ethernet68/1 (0.00s)
    --- PASS: TestGnmiGet/use_vendor_alias:_get_COUNTERS:Ethernet68/1_SAI_PORT_STAT_PFC_7_RX_PKTS (0.00s)
    --- PASS: TestGnmiGet/get_COUNTERS:Ethernet* (0.00s)
    --- PASS: TestGnmiGet/get_COUNTERS:Ethernet*_SAI_PORT_STAT_PFC_7_RX_PKTS (0.00s)
=== RUN   TestGnmiSubscribe
=== RUN   TestGnmiSubscribe/stream_query_for_table_COUNTERS_PORT_NAME_MAP_with_new_test_field_field
=== RUN   TestGnmiSubscribe/stream_query_for_table_key_Ethernet68_with_new_test_field_field
=== RUN   TestGnmiSubscribe/use_vendor_alias:_stream_query_for_table_key_Ethernet68/1_with_new_test_field_field
=== RUN   TestGnmiSubscribe/stream_query_for_COUNTERS/Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_with_update_of_filed_value
=== RUN   TestGnmiSubscribe/use_vendor_alias:_stream_query_for_COUNTERS/[Ethernet68/1]/SAI_PORT_STAT_PFC_7_RX_PKTS_with_update_of_filed_value
=== RUN   TestGnmiSubscribe/stream_query_for_table_key_Ethernet*_with_new_test_field_field_on_Ethernet68
=== RUN   TestGnmiSubscribe/stream_query_for_table_key_Ethernet*/SAI_PORT_STAT_PFC_7_RX_PKTS_with_field_value_update
=== RUN   TestGnmiSubscribe/poll_query_for_table_COUNTERS_PORT_NAME_MAP_with_new_field_test_field
=== RUN   TestGnmiSubscribe/poll_query_for_table_COUNTERS_PORT_NAME_MAP_with_test_field_delete
=== RUN   TestGnmiSubscribe/poll_query_for_COUNTERS/Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_with_field_value_change
=== RUN   TestGnmiSubscribe/use_vendor_alias:_poll_query_for_COUNTERS/[Ethernet68/1]/SAI_PORT_STAT_PFC_7_RX_PKTS_with_field_value_change
=== RUN   TestGnmiSubscribe/poll_query_for_table_key_Ethernet*_with_Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_field_value_change
=== RUN   TestGnmiSubscribe/poll_query_for_table_key_field_Ethernet*/SAI_PORT_STAT_PFC_7_RX_PKTS_with_Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_field_value_change
=== RUN   TestGnmiSubscribe/poll_query_for_COUNTERS/Ethernet*/Queues
=== RUN   TestGnmiSubscribe/poll_query_for_COUNTERS/Ethernet68/Queues_with_field_value_change
=== RUN   TestGnmiSubscribe/use_vendor_alias:_poll_query_for_COUNTERS/Ethernet68/Queues_with_field_value_change
--- PASS: TestGnmiSubscribe (56.04s)
    --- PASS: TestGnmiSubscribe/stream_query_for_table_COUNTERS_PORT_NAME_MAP_with_new_test_field_field (2.01s)
    --- PASS: TestGnmiSubscribe/stream_query_for_table_key_Ethernet68_with_new_test_field_field (3.01s)
    --- PASS: TestGnmiSubscribe/use_vendor_alias:_stream_query_for_table_key_Ethernet68/1_with_new_test_field_field (3.01s)
    --- PASS: TestGnmiSubscribe/stream_query_for_COUNTERS/Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_with_update_of_filed_value (3.01s)
    --- PASS: TestGnmiSubscribe/use_vendor_alias:_stream_query_for_COUNTERS/[Ethernet68/1]/SAI_PORT_STAT_PFC_7_RX_PKTS_with_update_of_filed_value (3.01s)
    --- PASS: TestGnmiSubscribe/stream_query_for_table_key_Ethernet*_with_new_test_field_field_on_Ethernet68 (3.02s)
    --- PASS: TestGnmiSubscribe/stream_query_for_table_key_Ethernet*/SAI_PORT_STAT_PFC_7_RX_PKTS_with_field_value_update (2.00s)
    --- PASS: TestGnmiSubscribe/poll_query_for_table_COUNTERS_PORT_NAME_MAP_with_new_field_test_field (2.02s)
    --- PASS: TestGnmiSubscribe/poll_query_for_table_COUNTERS_PORT_NAME_MAP_with_test_field_delete (2.02s)
    --- PASS: TestGnmiSubscribe/poll_query_for_COUNTERS/Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_with_field_value_change (2.01s)
    --- PASS: TestGnmiSubscribe/use_vendor_alias:_poll_query_for_COUNTERS/[Ethernet68/1]/SAI_PORT_STAT_PFC_7_RX_PKTS_with_field_value_change (2.01s)
    --- PASS: TestGnmiSubscribe/poll_query_for_table_key_Ethernet*_with_Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_field_value_change (2.08s)
    --- PASS: TestGnmiSubscribe/poll_query_for_table_key_field_Ethernet*/SAI_PORT_STAT_PFC_7_RX_PKTS_with_Ethernet68/SAI_PORT_STAT_PFC_7_RX_PKTS_field_value_change (2.05s)
    --- PASS: TestGnmiSubscribe/poll_query_for_COUNTERS/Ethernet*/Queues (1.11s)
    --- PASS: TestGnmiSubscribe/poll_query_for_COUNTERS/Ethernet68/Queues_with_field_value_change (2.01s)
    --- PASS: TestGnmiSubscribe/use_vendor_alias:_poll_query_for_COUNTERS/Ethernet68/Queues_with_field_value_change (2.03s)
PASS
ok      test/sonic-telemetry-backup/gnmi_server (cached)
```
-----------------------------------------------------------------

```
cchliu@cchliu-vm:~/Work/go/src/test/sonic-telemetry-backup$ go test -v ./dialout/dialout_client/
=== RUN   TestGNMIDialOutPublish
=== RUN   TestGNMIDialOutPublish/DialOut_to_first_collector_in_stream_mode_and_synced
=== RUN   TestGNMIDialOutPublish/DialOut_to_second_collector_in_stream_mode_upon_failure_of_first_collector
--- PASS: TestGNMIDialOutPublish (18.75s)
    --- PASS: TestGNMIDialOutPublish/DialOut_to_first_collector_in_stream_mode_and_synced (0.51s)
    --- PASS: TestGNMIDialOutPublish/DialOut_to_second_collector_in_stream_mode_upon_failure_of_first_collector (7.51s)
PASS
ok      test/sonic-telemetry-backup/dialout/dialout_client      (cached)

```

File: testdata/COUNTERS:Ethernet68:Queues_alias.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+{"Ethernet68/1:0":{},"Ethernet68/1:1":{"SAI_QUEUE_STAT_BYTES":"0","SAI_QUEUE_STAT_DROPPED_BYTES":"0","SAI_QUEUE_STAT_DROPPED_PACKETS":"0","SAI_QUEUE_STAT_PACKETS":"0"},"Ethernet68/1:10":{},"Ethernet68/1:11":{},"Ethernet68/1:12":{},"Ethernet68/1:13":{},"Ethernet68/1:14":{},"Ethernet68/1:15":{},"Ethernet68/1:16":{},"Ethernet68/1:17":{},"Ethernet68/1:18":{},"Ethernet68/1:19":{},"Ethernet68/1:2":{},"Ethernet68/1:3":{},"Ethernet68/1:4":{},"Ethernet68/1:5":{},"Ethernet68/1:6":{},"Ethernet68/1:7":{},"Ethernet68/1:8":{},"Ethernet68/1:9":{}}
```
Comment (lguohan) [inline testdata/COUNTERS:Ethernet68:Queues_alias.txt]: >{"Ethernet68/1:0":{},"Ethernet68/ [](start = 0, length = 33)

can you beautify them? hard to read in oneline.

File: sonic_data_client/virtual_db.go [hunk 2]
```diff
@@ -104,19 +153,34 @@ func v2rEthPortStats(paths []string) ([]tablePath, error) {
...
-				dbName:       paths[DbIdx],
-				tableName:    paths[TblIdx],
-				tableKey:     oid,
-				delimitor:    separator,
-				jsonTableKey: port,
+				dbName:    paths[DbIdx],
+				tableName: paths[TblIdx],
+				tableKey:  oid,
+				delimitor: separator,
+				//jsonTableKey: port,
```
Comment (lguohan) [inline sonic_data_client/virtual_db.go]: do not leave comments, do not change existing code spacing unless necessary.

File: sonic_data_client/virtual_db.go [hunk 3]
```diff
@@ -34,6 +34,11 @@ var (
 	// Queue name to oid map in COUNTERS table of COUNTERS_DB
 	countersQueueNameMap = make(map[string]string)
 
+	// Alias translation: from external name to internal name
+	e2i_aliasMap = make(map[string]string)
```
Comment (lguohan) [inline sonic_data_client/virtual_db.go]: >e2i_aliasMap [](start = 1, length = 12)

change to name2aliasMap and alias2nameMap.

Comment (lguohan) [reply sonic_data_client/virtual_db.go]: there is no internal/external name concept, let use name to refer to sonic interface name and alias to refer vendor port name.

---
In reply to: [213116801](https://github.com/Azure/sonic-telemetry/pull/13#discussion_r213116801) [](ancestors = 213116801)

File: sonic_data_client/virtual_db.go [hunk 4]
```diff
@@ -104,19 +153,34 @@ func v2rEthPortStats(paths []string) ([]tablePath, error) {
...
+				delimitor: separator,
+				//jsonTableKey: port,
+				jsonTableKey: oport,
 			}
 			tblPaths = append(tblPaths, tblPath)
 		}
 	} else { //single port
-		oid, ok := countersPortNameMap[paths[KeyIdx]]
+		var eport, iport string
+		iport = paths[KeyIdx]
```
Comment (lguohan) [inline sonic_data_client/virtual_db.go]: >iport [](start = 2, length = 5)

is this iport or alias?

File: sonic_data_client/virtual_db.go [hunk 5]
```diff
@@ -104,19 +153,34 @@ func v2rEthPortStats(paths []string) ([]tablePath, error) {
...
+		if val, ok := e2i_aliasMap[iport]; ok {
```
Comment (lguohan) [inline sonic_data_client/virtual_db.go]: >iport [](start = 29, length = 5)

alias

File: sonic_data_client/virtual_db.go [hunk 6]
```diff
@@ -140,21 +204,36 @@ func v2rEthPortFieldStats(paths []string) ([]tablePath, error) {
...
-				tableKey:     oid,
-				field:        paths[FieldIdx],
-				delimitor:    separator,
-				jsonTableKey: port,
+				dbName:    paths[DbIdx],
+				tableName: paths[TblIdx],
+				tableKey:  oid,
+				field:     paths[FieldIdx],
+				delimitor: separator,
+				//jsonTableKey: port,
```
Comment (lguohan) [inline sonic_data_client/virtual_db.go]: remove comments.

File: sonic_data_client/virtual_db.go (hunk 2 above)
Comment (jleveque) [reply sonic_data_client/virtual_db.go]: I believe the changes in spacing are due to the use of "gofmt", which unified the code to the standards of the Go language. I believe we should keep these changes and continue using gofmt to ensure consistent style going forward.

Comment (chaliu) [reply sonic_data_client/virtual_db.go]: Agree. The changes are made due to the use of "gofmt", which targets to unify code formats.

File: sonic_data_client/virtual_db.go (hunk 3 above)
Comment (chaliu) [reply sonic_data_client/virtual_db.go]: Variable names are renamed. Please review.

File: testdata/COUNTERS:Ethernet68:Queues_alias.txt (hunk 1 above)
Comment (chaliu) [reply testdata/COUNTERS:Ethernet68:Queues_alias.txt]: JSON files are pretty-printed.

File: sonic_data_client/virtual_db.go (hunk 2 above)
Comment (jleveque) [reply sonic_data_client/virtual_db.go]: Is there a good reason to keep the commented-out line? If not, please remove.

Comment (chaliu) [reply sonic_data_client/virtual_db.go]: It was for me to track the changed place. Can remove it easily. Other comments?

File: sonic_data_client/virtual_db.go (hunk 6 above)
Comment (jleveque) [reply sonic_data_client/virtual_db.go]: Again, unless there is there a good reason to keep the commented-out line, please remove.

File: sonic_data_client/virtual_db.go [hunk 7]
```diff
@@ -104,19 +153,33 @@ func v2rEthPortStats(paths []string) ([]tablePath, error) {
...
+		var alias, name string
+		alias = paths[KeyIdx]
+		name = alias
+		if val, ok := alias2nameMap[alias]; ok {
+			name = val
+		}
+		oid, ok := countersPortNameMap[name]
 		if !ok {
-			return nil, fmt.Errorf(" %v not a valid port ", paths[KeyIdx])
+			return nil, fmt.Errorf("%v not a valid sonic interface port, vendor alias is %v", name, alias)
```
Comment (jleveque) [inline sonic_data_client/virtual_db.go]: Suggest rewording to remove redundancy: `"%v not a valid SONiC interface. Vendor alias is %v"`

File: sonic_data_client/virtual_db.go [hunk 8]
```diff
@@ -140,21 +203,35 @@ func v2rEthPortFieldStats(paths []string) ([]tablePath, error) {
...
+		var alias, name string
+		alias = paths[KeyIdx]
+		name = alias
+		if val, ok := alias2nameMap[alias]; ok {
+			name = val
+		}
+		oid, ok := countersPortNameMap[name]
 		if !ok {
-			return nil, fmt.Errorf(" %v not a valid port ", paths[KeyIdx])
+			return nil, fmt.Errorf(" %v not a valid sonic interface port, vendor alias is %v ", name, alias)
```
Comment (jleveque) [inline sonic_data_client/virtual_db.go]: Suggest rewording to remove redundancy: `"%v not a valid SONiC interface. Vendor alias is %v"`
